### PR TITLE
Install docker in the OpenHands app image

### PR DESCRIPTION
**Pull request by mamoodi**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

Install docker in the OpenHands app image

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

Installing docker in the OpenHands app image allows it to build a sandbox image off of a SANDBOX_BASE_CONTAINER_IMAGE using the docker method. So users are able to provide a base image to the docker command without the development flow.

I tested this by:
* Building the OpenHands app image and tagging it with `myapp:1.0.0`
* Running the docker command as follows:
```
docker run -it \                                         
    -e SANDBOX_BASE_CONTAINER_IMAGE=ruby:3.3.4 \
    -e SANDBOX_USER_ID=$(id -u) \
    -e WORKSPACE_MOUNT_PATH=$WORKSPACE_BASE \
    -v $WORKSPACE_BASE:/opt/workspace_base \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -p 3000:3000 \
    --add-host host.docker.internal:host-gateway \
    --name openhands-app-$(date +%Y%m%d%H%M%S) \
    myapp:1.0.0 
```
* Checking that ruby 3.3.4 is installed

**I am unsure if this is the route we want to go. It seems to add 0.5GB to the app image. However, this is what I came up with after some guidance. See the issues for why this change is needed below. If this is fine, I can this method to the https://docs.all-hands.dev/modules/usage/how-to/custom-sandbox-guide doc**

---
**Link of any specific issues this addresses**
https://github.com/All-Hands-AI/OpenHands/issues/4220
https://github.com/All-Hands-AI/OpenHands/issues/4230

### CODE REVIEW

File: containers/app/Dockerfile [hunk 1]
```diff
@@ -46,6 +46,13 @@ RUN mkdir -p $WORKSPACE_BASE
 RUN apt-get update -y \
     && apt-get install -y curl ssh sudo
 
+# Install Docker
+RUN apt install -y apt-transport-https ca-certificates gnupg \
+    && curl -fsSL https://download.docker.com/linux/debian/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker.gpg \
+    && echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker.gpg] https://download.docker.com/linux/debian bookworm stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null \
+    && apt update \
```
Comment (tobitege) [inline containers/app/Dockerfile:53]: is `apt-update` needed twice?

Comment (mamoodi) [reply containers/app/Dockerfile:53]: I was following this: https://linuxiac.com/how-to-install-docker-on-debian-12-bookworm/
But I can remove those and see how it goes. The `apt update` may be needed after we add that source but I'm not linux expert lol. Will try a few things

Comment (mamoodi) [reply containers/app/Dockerfile]: @tobitege followed the official docs now: https://docs.docker.com/engine/install/debian/

**Comment by tobitege** [comment]
Have you taken a look at this (recent contribution)? Looks very similar to your steps, maybe that helps:
https://github.com/All-Hands-AI/OpenHands/blob/main/containers/dev/Dockerfile

**Comment by mamoodi** [comment]
> Have you taken a look at this (recent contribution)? Looks very similar to your steps, maybe that helps: https://github.com/All-Hands-AI/OpenHands/blob/main/containers/dev/Dockerfile

That one seems to be installing the Ubuntu one: https://docs.docker.com/engine/install/ubuntu/
I checked the app version and it's based off of Debian:
```
Distributor ID: Debian
Description:    Debian GNU/Linux 12 (bookworm)
Release:        12
Codename:       bookworm
```

So I followed that one.

**Review by xingyaoww — COMMENTED** [review]
This is great! Actually - can we add a quick test for this workflow? e.g., try to run `docker run hello-world` inside the APP image?

File: containers/app/Dockerfile (hunk 1 above)
Comment (rbren) [reply containers/app/Dockerfile]: Yup after you add a new source you need to update

**Review by rbren — APPROVED** [review]

**Comment by rbren** [comment]
test would definitely be nice to have!

**Comment by mamoodi** [comment]
@rbren and @xingyaoww added test based on Xingyao's suggestion.

**Pull request merged**